pkg/storage/unified/indexing/main.go

```diff
@@ -0,0 +1,340 @@
+		//err := index.Object(hit.ID).VisitFields(func(field bleve.DocumentField) {
+		//	switch field.Name() {
+		//	case "Title":
+		//		doc.Title = string(field.Value())
+		//	case "Content":
+		//		doc.Content = string(field.Value())
+		//	case "Tags":
+		//		doc.Tags = append(doc.Tags, string(field.Value()))
+		//	}
+		//})
+		//if err != nil {
+		//	log.Printf("Failed to retrieve document %s: %v", hit.ID, err)
+		//	continue
+		//}
+		//
+		//fmt.Printf("Title: %s\n", doc.Title)
+		//fmt.Printf("Content: %s\n", doc.Content)
+		//fmt.Printf("Tags: %v\n", doc.Tags)
+		//fmt.Println("---------")
+	}
+}
+
+func seedObjects() []Object {
+	type Resource struct {
+		Group      string
+		Resource   string
+		Kind       string
+		ApiVersion string
+	}
+
+	namespaces := []string{"stack-1", "stack-2"}
+	resources := []Resource{
+		{
+			Group:      "playlist.grafana.app",
+			Resource:   "playlists",
+			Kind:       "Playlist",
+			ApiVersion: "playlist.grafana.app/v0alpha1",
+		},
+		{
+			Group:      "folder.grafana.app",
+			Resource:   "folders",
+			Kind:       "Folder",
+			ApiVersion: "folder.grafana.app/v0alpha1",
+		},
+	}
+
+	// creates 4,000,000 objects
+	objects := []Object{}
+	for _, namespace := range namespaces {
+		for _, resource := range resources {
+			for i := 0; i < 10000; i++ {
+				objects = append(objects, Object{
+					Guid:            fmt.Sprintf("%s-%s-%d", namespace, resource.Resource, i),
+					ResourceVersion: time.Now().UnixNano(),
+					Group:           resource.Group,
+					Resource:        resource.Resource,
+					Namespace:       namespace,
+					Name:            fmt.Sprintf("%s-%s-%d", namespace, resource.Resource, i),
+					Value: K8sObject{
+						Kind:       resource.Kind,
+						ApiVersion: resource.ApiVersion,
+						K8sMeta: K8sMeta{
+							Name:              fmt.Sprintf("%s-%s-%d", namespace, resource.Resource, i),
+							Namespace:         namespace,
+							Uid:               fmt.Sprintf("k8s-uid-%d", i),
+							CreationTimestamp: time.Now().Add(-time.Duration(i) * time.Hour).Format(time.RFC3339),
+						},
+					},
+				})
+			}
+		}
+	}
+
+	return objects
+}
+
+func IndexObjects(index bleve.Index, objects []Object) {
+	for _, obj := range objects {
+		err := index.Index(obj.Guid, obj)
+		if err != nil {
+			fmt.Println("Failed to index document:", err)
+		}
+	}
+
+	fmt.Println("Indexing synchronously completed")
+}
+
+func indexObjectsConcurrently(index bleve.Index, objects []Object) {
+	// Create a wait group to wait for all goroutines to finish
+	var wg sync.WaitGroup
+
+	// Create a channel to send batches to the workers
+	batchChan := make(chan *bleve.Batch)
+
+	// Start workers (goroutines)
+	for i := 0; i < INDEX_WORKERS; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for batch := range batchChan {
+				err := index.Batch(batch)
+				if err != nil {
+					log.Fatal(err)
+				}
+			}
+		}()
+	}
+
+	// Create and send batches
+	for i := 0; i < len(objects); i += INDEX_BATCH_SIZE {
```

Comment:
hmm i think this can be optimized.  curious to see how long it takes to make the batches.  seems this is not really concurrent since it takes a bit of time to loop over and fill the batch before sending it to the channel.  maybe would be better to just send the chunk of objects to a channel, which will create the batch and fill it

Reply:
i tried it.  it was taking about 5 seconds to create the batches.  moving the batch creating into the go routine sped that up but overall shaved off only about 1 second.  i think boltdb must lock during a batch append